### ISSUE TITLE
在V100 GPU上训练Qwen2.5-7B时，观察到lr, grad_norm, reward_std等指标在训练全程持续为nan，数据类型使用float16。请教是否是因为数据类型导致这个问题？

### ISSUE THREAD

**Issue by Dielianss** [lzhxmu/CPPO#3]
使用本项目代码复现实验时，在训练开始时，计算reward_std出现如下警告：
UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1823.)
  std_grouped_rewards = rewards.view(-1, self.num_generations * self.repeat ).std(dim=1)

随后在训练全程，lr, grad_norm, reward_std等指标持续为nan，展示如下：

{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 250.62500762939453, 'rewards/xmlcount_reward_func': -0.3253333419561386, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re
ward_func': 0.0, 'rewards/int_reward_func': 0.06250000186264515, 'rewards/correctness_reward_func': 0.0, 'reward': 0.21633333899080753, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'k
l': 0.0, 'clip_ratio': 0.0, 'epoch': 2.12}
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13}
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 273.75000762939453, 'rewards/xmlcount_reward_func': -0.3218333385884762, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re
ward_func': 0.0, 'rewards/int_reward_func': 0.10416666977107525, 'rewards/correctness_reward_func': 0.0, 'reward': 0.26150000281631947, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'k
l': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13}
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13}
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 218.00000762939453, 'rewards/xmlcount_reward_func': -0.2538333311676979, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re
ward_func': 0.0, 'rewards/int_reward_func': 0.0, 'rewards/correctness_reward_func': 0.0, 'reward': 0.2253333404660225, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ra
tio': 0.0, 'epoch': 2.13}

lr尝试过5e-6/5e-5这两组，结果都一样。
数据类型选择：float16。
查询网上资料，可能是由于V100太老，不支持bfloat16，需要向上使用float32才可以，请问是否是这样？感谢！

**Comment by lzhxmu** [comment]
可以试试看 BF16

**Comment by lzhxmu** [comment]
> 使用本项目代码复现实验时，在训练开始时，计算reward_std出现如下警告： UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1823.) std_grouped_rewards = rewards.view(-1, self.num_generations * self.repeat ).std(dim=1)
> 
> 随后在训练全程，lr, grad_norm, reward_std等指标持续为nan，展示如下：
> 
> {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 250.62500762939453, 'rewards/xmlcount_reward_func': -0.3253333419561386, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re ward_func': 0.0, 'rewards/int_reward_func': 0.06250000186264515, 'rewards/correctness_reward_func': 0.0, 'reward': 0.21633333899080753, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'k l': 0.0, 'clip_ratio': 0.0, 'epoch': 2.12} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 273.75000762939453, 'rewards/xmlcount_reward_func': -0.3218333385884762, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re ward_func': 0.0, 'rewards/int_reward_func': 0.10416666977107525, 'rewards/correctness_reward_func': 0.0, 'reward': 0.26150000281631947, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'k l': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 218.00000762939453, 'rewards/xmlcount_reward_func': -0.2538333311676979, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re ward_func': 0.0, 'rewards/int_reward_func': 0.0, 'rewards/correctness_reward_func': 0.0, 'reward': 0.2253333404660225, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ra tio': 0.0, 'epoch': 2.13}
> 
> lr尝试过5e-6/5e-5这两组，结果都一样。 数据类型选择：float16。 查询网上资料，可能是由于V100太老，不支持bfloat16，需要向上使用float32才可以，请问是否是这样？感谢！

很有可能是 fp16 的原因，建议使用 fp32/ bf16

**Comment by Dielianss** [comment]
> > 使用本项目代码复现实验时，在训练开始时，计算reward_std出现如下警告： UserWarning: std(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1823.) std_grouped_rewards = rewards.view(-1, self.num_generations * self.repeat ).std(dim=1)
> > 随后在训练全程，lr, grad_norm, reward_std等指标持续为nan，展示如下：
> > {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 250.62500762939453, 'rewards/xmlcount_reward_func': -0.3253333419561386, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re ward_func': 0.0, 'rewards/int_reward_func': 0.06250000186264515, 'rewards/correctness_reward_func': 0.0, 'reward': 0.21633333899080753, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'k l': 0.0, 'clip_ratio': 0.0, 'epoch': 2.12} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 273.75000762939453, 'rewards/xmlcount_reward_func': -0.3218333385884762, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re ward_func': 0.0, 'rewards/int_reward_func': 0.10416666977107525, 'rewards/correctness_reward_func': 0.0, 'reward': 0.26150000281631947, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'k l': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ratio': 0.0, 'epoch': 2.13} {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.0, 'pruning_rate': 0.5, 'completion_length': 218.00000762939453, 'rewards/xmlcount_reward_func': -0.2538333311676979, 'rewards/soft_format_reward_func': 0.4791666716337204, 'rewards/strict_format_re ward_func': 0.0, 'rewards/int_reward_func': 0.0, 'rewards/correctness_reward_func': 0.0, 'reward': 0.2253333404660225, 'reward_std': nan, 'second_item': 0.0, 'first_item': nan, 'first_item_div_second_item': nan, 'total_sum': nan, 'kl': 0.0, 'clip_ra tio': 0.0, 'epoch': 2.13}
> > lr尝试过5e-6/5e-5这两组，结果都一样。 数据类型选择：float16。 查询网上资料，可能是由于V100太老，不支持bfloat16，需要向上使用float32才可以，请问是否是这样？感谢！
> 
> 很有可能是 fp16 的原因，建议使用 fp32/ bf16

好的，感谢
